R/mod_hierarchical_count_table.R

```diff
@@ -43,135 +200,219 @@ EC <- poc( # nolint
+      warning_message <- EC$MSG$VALIDATE$ORIG_AFTER_CENSOR
+      subset_event_df <- subset_event_df[0, ]
+    }
+  }
 
-    curr_count_df <- dplyr::bind_rows(curr_grouped_count_df, curr_total_count_df)
-    curr_count_df[, empty_hierarchy] <- factor(.special_char)
-    curr_count_df[, hier_lvl_col] <- idx
+  # Remove rows with missing event dates
+  if (has_event_dt) subset_event_df <- subset_event_df[!is.na(subset_event_df[[event_date_var]]), ]
 
-    hierarchy_df_list[[idx]] <- curr_count_df
+  # Remove events that occur before origin date (implicitly when origin date is missing)
+  if (has_origin_dt && has_event_dt) {
+    subset_event_df <- subset_event_df |>
+      dplyr::left_join(subset_pop_df[, c(subjid_var, origin_date_var)], by = subjid_var) |>
+      dplyr::filter(.data[[event_date_var]] >= .data[[origin_date_var]]) |>
+      dplyr::select(-dplyr::all_of(origin_date_var))
   }
 
-  hierarchy_df <- dplyr::bind_rows(hierarchy_df_list)
+  # Remove events that occur after non-missing censor date
+  if (has_censor_dt && has_event_dt) {
+    subset_event_df <- subset_event_df |>
+      dplyr::left_join(subset_pop_df[, c(subjid_var, censor_date_var)], by = subjid_var) |>
+      dplyr::filter(is.na(.data[[censor_date_var]]) | .data[[event_date_var]] <= .data[[censor_date_var]]) |>
+      dplyr::select(-dplyr::all_of(censor_date_var))
+  }
+
+  # Prepare subject level analysis data ----
+
+  adtte <- create_adtte(event_df = subset_event_df,
+                        pop_df = subset_pop_df,
+                        hierarchy = hierarchy,
+                        group_var = group_var,
+                        subjid_var = subjid_var,
+                        origin_date_var = origin_date_var,
+                        censor_date_var = censor_date_var,
+                        event_date_var = event_date_var)
+
+  # Drop rows where time at risk could not be determined
+  if (compute_risk) {
+    invalid_rows <- which(is.na(adtte[["AVAL"]]))
+    if (length(invalid_rows) > 0) {
+      invalid_subjects <- unique(adtte[[subjid_var]][invalid_rows])
+      log_inform(paste("Time at risk could not be determined for the following subjects:",
+                       paste0('"', invalid_subjects, '"', collapse = ", ")), level = "inform")
+      adtte <- adtte[-invalid_rows, ]
+    }
+  }
+
+  # Add group totals ----
+
+  if (total) {
+    adtte_totals <- adtte
+    adtte_totals[[group_var]] <- total_group_val
+
+    adtte <- rbind(adtte, adtte_totals)
+  } else {
+    # No group total requested so clear the total group value
+    total_group_val <- character()
+  }
+
+  # Calculate statistics ----
+
+  calc_stats <- adtte |>
+    dplyr::mutate(dplyr::across(dplyr::all_of(hierarchy),
+                                ~ as.factor(replace(as.character(.), is.na(.), EC$VAL$SPECIAL_CHAR)))) |>
+
+    dplyr::group_by(dplyr::across(dplyr::all_of(c(hierarchy, group_var, hier_lvl_col))))
+
+  if (compute_risk && "AVAL" %in% names(adtte)) {
```

Comment:
When could it ever be the case that `AVAL` is not in `adtte` if `compute_risk` is true? In line 332 you access it but you make no check. Given that you are in charge of creating `adtte` I would consider this checking if `AVAL` is a column in `adtte` a developer assert, but not a real case that could happen in the app. Therefore I would not include it in the if condition, but maybe in a `stopifnot` or similar, maybe with a comment indicating its developer nature. In some other codebases we have a function like

```r
assert <- function(expr, msg) {
  if (!isTRUE(expr)) stop(msg, call. = FALSE)
}
```

This way we differentiate real possible situations from developer asserts indicating impossible situations that indicate errors in the code.